Gemfile

```diff
@@ -43,6 +46,9 @@ group :test do
   gem "capybara", ">= 2.15"
   gem "selenium-webdriver"
   gem "chromedriver-helper"
+
+  gem "shoulda-matchers", "~> 3.1"
+  gem "database_cleaner"
```

Comment:
gem này dùng làm gì?

Reply:
- Gem Shoulda Matchers cung cấp cách viết 1 dòng lệnh Rspec dùng để test các chức năng của Rails, giúp viết test 1 cách ngắn gọn, rõ ràng và ít gặp lỗi hơn.
Shoulda Matchers cùng cấp cho ta rất nhiều phương thức dùng để test .
Một số method thường dùng:
+ ActiveModel:
have_secure_password
validate_absence_of
validate_acceptance_of
validate_confirmation_of
validate_exclusion_of
validate_inclusion_of
validate_length_of
validate_numericality_of
validate_presence_of
accept_nested_attributes_for
belong_to
define_enum_for
have_and_belong_to_many
have_db_column
have_db_index tests
have_many
have_one
have_readonly_attribute
validate_uniqueness_of
ví dụ: it {is_expected.to validate_presence_of :name}
+ ActionController : Controller:
filter_param
permit tests
redirect_to
render_template
render_with_layout
rescue_from tests
respond_with tests
route
set_flash

- Gem database_cleaner:  đảm bảo việc database luôn nhất quán trong suốt quá trình chạy test, tức là sau khi chạy xong một test nó sẽ rollback database lại trạng thái trước khi chạy test, đảm bảo database luôn nhất quán trong suốt quá trình thực hiện tất cả các test case.